File: src/Ch3/Money.java
package Ch3;

import java.util.Currency;

public class Money {
    final int amount;
    final Currency currency;

    Money(final int amount, final Currency currency) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount is not more than 0.");
        }
        if (currency == null) {
            throw new NullPointerException("Please specify currency");
        }

        this.amount = amount;
        this.currency = currency;
    }

    Money add(final Money other) {
        if (!currency.equals(other.currency)) {
            throw new IllegalArgumentException("Currency units are different.");
        }

        final int added = amount + other.amount;
        return new Money(added, currency);
    }


}

File: src/Ch3/money.rb
class Money
  def initialize(amount, currency)
    raise ArgumentError('amount is not more than 0.') if amount.negative?
    raise ArgumentError('currency is not designated correctly.') if currency.blank?

    @amount = amount
    @currency = currency
    freeze
  end

  def add(other)
    raise ArgumentError('currecy units are different.') if @currency != other.currency

    added = @amount + other.amount
    Money.new(added, @currency)
  end
end
